drivers/serial/uart_sam.c

```diff
@@ -364,206 +364,74 @@ static const struct uart_driver_api uart_sam_driver_api = {
 #endif	/* CONFIG_UART_INTERRUPT_DRIVEN */
 };
 
-/* UART0 */
-
-#ifdef CONFIG_UART_SAM_PORT_0
-
 #ifdef CONFIG_UART_INTERRUPT_DRIVEN
-/* Forward declare function */
-static void uart0_sam_irq_config_func(struct device *port);
-#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 
-static const struct uart_sam_dev_cfg uart0_sam_config = {
-	.regs = UART0,
-	.periph_id = ID_UART0,
-	.pin_rx = PIN_UART0_RXD,
-	.pin_tx = PIN_UART0_TXD,
+#define UART_SAM_IRQ_CALLBACK(n)					\
+	.irq_config_func = uart##n##_sam_irq_config_func,
 
-#ifdef CONFIG_UART_INTERRUPT_DRIVEN
-	.irq_config_func = uart0_sam_irq_config_func,
-#endif
-};
+#define UART_SAM_IRQ_CONNECT(n)					\
+do {									\
+	IRQ_CONNECT(DT_IRQN(DT_NODELABEL(uart##n)),			\
+		    DT_IRQ(DT_NODELABEL(uart##n, priority)),		\
+		    uart_sam_isr, DEVICE_GET(uart##n##_sam), 0);	\
+	irq_enable(DT_IRQN(DT_NODELABEL(uart##n)));			\
+} while (0)
 
-static struct uart_sam_dev_data uart0_sam_data = {
-	.baud_rate = DT_UART_SAM_PORT_0_BAUD_RATE,
-};
-
-DEVICE_AND_API_INIT(uart0_sam, DT_UART_SAM_PORT_0_NAME, &uart_sam_init,
-		    &uart0_sam_data, &uart0_sam_config, PRE_KERNEL_1,
-		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &uart_sam_driver_api);
+#define UART_SAM_IRQ_HANDLER_DECL(n)					\
+static void uart##n##_sam_irq_config_func(struct device *port);
 
-#ifdef CONFIG_UART_INTERRUPT_DRIVEN
-static void uart0_sam_irq_config_func(struct device *port)
-{
-	IRQ_CONNECT(DT_UART_SAM_PORT_0_IRQ,
-		    DT_UART_SAM_PORT_0_IRQ_PRIO,
-		    uart_sam_isr,
-		    DEVICE_GET(uart0_sam), 0);
-	irq_enable(DT_UART_SAM_PORT_0_IRQ);
+#define UART_SAM_IRQ_HANDLER(n)					\
+static void uart##n##_sam_irq_config_func(struct device *port)		\
+{									\
+	UART_SAM_IRQ_CONNECT(n);					\
 }
-#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
-#endif /* CONFIG_UART_SAM_PORT_0 */
 
-/* UART1 */
+#else
 
-#ifdef CONFIG_UART_SAM_PORT_1
+#define UART_SAM_IRQ_CALLBACK(n)
+#define UART_SAM_IRQ_HANDLER_DECL(n)
+#define UART_SAM_IRQ_HANDLER(n)
 
-#ifdef CONFIG_UART_INTERRUPT_DRIVEN
-/* Forward declare function */
-static void uart1_sam_irq_config_func(struct device *port);
 #endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 
-static const struct uart_sam_dev_cfg uart1_sam_config = {
-	.regs = UART1,
-	.periph_id = ID_UART1,
-	.pin_rx = PIN_UART1_RXD,
-	.pin_tx = PIN_UART1_TXD,
-
-#ifdef CONFIG_UART_INTERRUPT_DRIVEN
-	.irq_config_func = uart1_sam_irq_config_func,
+#define UART_SAM_DEVICE_INIT(n)					\
+UART_SAM_IRQ_HANDLER_DECL(n)						\
+									\
+static const struct uart_sam_dev_cfg uart##n##_sam_config = {		\
+	.regs = (Uart *)DT_REG_ADDR(DT_NODELABEL(uart##n)),		\
+	.periph_id = DT_PROP(DT_NODELABEL(uart##n), peripheral_id),	\
+	.pin_rx = PIN_UART##n##_RXD,					\
+	.pin_tx = PIN_UART##n##_TXD,					\
+	UART_SAM_IRQ_CALLBACK(n)					\
```

Comment:
On a somewhat unrelated note, as static device tree support grows more complex and advanced, I wonder if we should start considering the introduction of an external preprocessor/codegen for this purpose (e.g. using m4, or modern equivalent).

Working with these multi-line C preprocessor macros is hard enough and using `#if/#ifdef`s inside one is impossible.

cc @galak @mbolivar-nordic

Reply:
I am not yet convinced this is necessary and I would like to avoid it if possible. Having Kconfig and DT is already a big learning curve for traditional bare metal embedded developers. Adding another meta/config language to Zephyr proper is a big step. Nothing is stopping out of tree driver writers from doing what they want, of course.